File: applications/search/lib/IndexClient.py
import json
import urllib2
import base64

class Document(object):

    def __init__(self, index_client):
        self.index_client = index_client
        self.identifier = None
        self.collection = None
        self.metas = {}
        self.fields = {}

    def set_identifier(self, identifier):
        self.identifier = identifier
        return self

    def set_collection(self, collection):
        self.collection = collection
        return self

    def get_collection(self):
        return self.collection

    def get_identifier(self):
        return self.identifier

    def get_meta(self, meta_name):
        if meta_name in self.metas:
            return self.metas[meta_name]

    def set_meta(self, meta_name, meta_value):
        self.metas[meta_name] = meta_value
        return self

    def add_field(self, field_name, field_value, analyzer = "standard"):
        self.fields[field_name] = {'value':field_value, 'analyzer':analyzer}
        return self

    def add_to_index(self):
        to_json = {
            'action':'index',
            'setIdentifier':self.identifier,
            'setCollection':self.collection,
            'setMetas':self.metas,
            'addFields':self.fields}
        self.index_client.post_request(json.dumps(to_json))
        return self
        
class Search(object):

    def __init__(self, index_client):
        self.index_client = index_client
        self.total = 0
        self.type = 'and'
        self.collection = None
        self.field = None
        self.analyze = []
        self.offset = 0
        self.lmit = 20
        self.analyzed = []

    def set_type(self,type = 'and'):
        if type not in ['and','or']:
            raise IndexException('unknown search type')
        self.type = type
        return self

    def limit_to_collection(self,collection):
        self.collection = collection
        return self

    def limit_to_field(self, field):
        self.field = field
        return self

    def analyze_text(self, text, analyzer = 'standard'):
        self.analyze.append({'value':text,'analyzer':analyzer})
        return self

    def get_total(self):
        return self.total

    def get_analyzed_words(self):
        return self.analyzed

    def get_documents(self, offset = 0, limit = 20):
        self.offset = offset
        self.limit = limit
        to_json = {
            'action':'search',
            'setType':self.type,
            'limitToCollection':self.collection,
            'limitToField':self.field,
            'analyzeText':self.analyze,
            'getDocuments':{'offset':self.offset,'limit':self.limit}}
        results = self.index_client.post_request(json.dumps(to_json))
        self.total = results['total']
        self.analyzed = results['analyzed']
        returned_results = []
        for raw_document in results['documents']:
            document = Document(self.index_client)
            document.set_collection(raw_document['collection'])
            document.set_identifier(raw_document['identifier'])
            for key in raw_document['metas'].keys():
                document.set_meta(key,raw_document['metas'][key])
            returned_results.append(document)
        return returned_results

class IndexException(Exception):

    def __init__(self, message):
        self.message = message

    def __str__(self):
        return repr(self.message)

class IndexClient(object):

    def __init__(self, endpoint, login, password):
        self.endpoint = endpoint
        self.login = login
        self.password = password

    def set_endpoint(self, endpoint):
        self.endpoint = endpoint
        return self

    def set_login(self, login):
        self.login = login
        return self

    def set_password(self, password):
        self.password = password
        return self

    def delete_document(self, identifier):
        self.post_request(json.dumps({'action':'delete', 'identifier':identifier}))
        return self

    def new_document(self):
        return Document(self)
    
    def new_search(self):
        return Search(self)

    def post_request(self, json_string):
        response_string = self.file_post_contents(self.endpoint, json_string)
        if response_string == False:
            raise IndexException('could not get response')
        response = json.loads(response_string)
        if response['status'] != "ok":
            raise IndexException(response['message'])
        return response

    def file_post_contents(self, url, data):
        req = urllib2.Request(url)
        auth_string = base64.encodestring('%s:%s' % (self.login, self.password))[:-1]
        req.add_header("Authorization", "Basic %s" % auth_string)
        req.add_header("Content-type", "application/json")
        f = urllib2.urlopen(req, data)
        return f.read()


if __name__ == "__main__":
    index = IndexClient('http://www.desfrenes.com/search/endpoint/', 'searchtest', 'boogaloo')
    document = index.new_document()
    document.set_collection('python client test')
    document.set_identifier('CDPYTHON')
    document.set_meta('url', 'http://www.nothere.com/test/')
    document.add_field('artist', 'guido')
    document.add_field('album', 'python in 3 hours')
    document.add_to_index()

    search = index.new_search()
    search.analyze_text('guido')
    results = search.get_documents()
    for document in results:
        print document.get_meta('url')

File: applications/search/lib/IndexClient.php
<?php
namespace search\lib;
class IndexClient
{
    private $endpoint;
    private $login;
    private $password;

    public function __construct($endpoint, $login, $password)
    {
        $this->endpoint = $endpoint;
        $this->login = $login;
        $this->password = $password;
    }

    public function setEndpoint($endpoint)
    {
        $this->endpoint = $endpoint;
        return $this;
    }

    public function setLogin($login)
    {
        $this->login = $login;
        return $this;
    }

    public function deleteDocument($identifier)
    {
        $toJSON = array(
            'action' => 'delete',
            'identifier' => $identifier
        );
        $this->postRequest(json_encode($toJSON));
        return $this;
    }

    public function setPassword($password)
    {
        $this->password = $password;
        return $this;
    }

    public function newDocument()
    {
        return new Document($this);
    }

    public function newSearch()
    {
        return new Search($this);
    }

    public function postRequest($json)
    {
        $response_string = $this->filePostContents($this->endpoint, $json, false);
        if($response_string === false)
        {
            throw new Exception('could not get response');
        }
        $response = json_decode($response_string, true);
        if($response === NULL)
        {
            $e = new Exception('could not decode response');
            $e->setLastRawResponse($response_string);
            throw $e;
        }
        
        if($response['status'] != 'ok')
        {
            throw new Exception($response['message']);
        }
        return $response;
    }


    private function filePostContents($url, $data, $headers = false)
    {
        $url = parse_url($url);
        if(is_array($data))
        {
            $_data = array();
            while(list($n,$v) = each($data))
            {
                $_data[] = "$n=".str_replace(array('/',':'),array('%2F','%3A'),$v);
            }
            $data = implode('&', $_data);
        }
        if (!isset($url['port']))
        {
            if ($url['scheme'] == 'http')
            {
                $url['port'] = 80;
            }
            elseif ($url['scheme'] == 'https')
            {
                $url['port'] = 443;
            }
        }
        $url['query'] = isset($url['query']) ? $url['query'] : '';

        $url['protocol'] = $url['scheme'] . '://';
        $eol = "\r\n";

        $post_content = 'POST ' . $url['protocol'] . $url['host'] . $url['path']. ' HTTP/1.0' . $eol.
            'Host: ' . $url['host'] . $eol.
            'Referer: ' . $url['protocol'] . $url['host'] . $url['path'] . $eol.
            'Content-Type: application/x-www-form-urlencoded' . $eol.
            'Content-Length: ' . strlen($data) . $eol.
            'Authorization: Basic '.base64_encode($this->login.':'.$this->password).$eol.
            $eol.$data;
        $fp = fsockopen(($url['port'] == '443' ? 'ssl://': '').$url['host'], $url['port'], $errno, $errstr, 30);
        if($fp)
        {
            fputs($fp, $post_content);
            $result = '';
            while(!feof($fp))
            {
                $result .= fgets($fp, 128);
            }
            fclose($fp);
            if (!$headers)
            {
                //removes headers
                $pattern="/^.*\r\n\r\n/s";
                $result=preg_replace($pattern,'',$result);
            }
            return $result;
        }
        return false;
    }
}

class Document
{
    private $index;
    private $identifier;
    private $collection;
    private $metas = array();
    private $fields = array();


    public function  __construct(IndexClient $index)
    {
        $this->index = $index;
    }

    public function setIdentifier($identifier)
    {
        $this->identifier = $identifier;
        return $this;
    }

    public function setCollection($collection)
    {
        $this->collection = $collection;
        return $this;
    }


    public function getCollection()
    {
        return $this->collection;
    }

    public function getIdentifier()
    {
        return $this->identifier;
    }

    public function getMeta($meta_name)
    {
        if(isset($this->metas[$meta_name]))
        {
            return $this->metas[$meta_name];
        }
    }

    /**
     *
     * @param string $meta_name
     * @param mixed $meta_value
     * @return Document
     */
    public function setMeta($meta_name, $meta_value)
    {
        $this->metas[$meta_name] = $meta_value;
        return $this;
    }

    public function addField($field_name, $field_value, $analyzer = 'standard')
    {
        $this->fields[$field_name] = array('value' => $field_value, 'analyzer' => $analyzer);
        return $this;
    }

    public function addToIndex()
    {
        $toJSON = array(
            'action' => 'index',
            'setIdentifier' => $this->identifier,
            'setCollection' => $this->collection,
            'setMetas' => $this->metas,
            'addFields' => $this->fields
        );
        $this->index->postRequest(json_encode($toJSON));
        return $this;
    }
}

class Search
{
    private $total = 0;
    private $type = 'and';
    private $collection;
    private $field;
    private $analyze = array();
    private $offset = 0;
    private $limit = 20;
    private $index;
    private $analyzed = array();

    public function  __construct(IndexClient $index)
    {
        $this->index = $index;
    }

    public function setType($type = 'and')
    {
        if(!in_array($type, array('and', 'or')))
        {
            throw new Exception('unknown search type "'.$type.'"');
        }
        $this->type = $type;
        return $this;
    }

    public function limitToCollection($collection)
    {
        $this->collection = $collection;
        return $this;
    }

    public function limitToField($field_name)
    {
        $this->field = $field_name;
        return $this;
    }

    public function analyzeText($text, $analyzer = 'standard')
    {
        $this->analyze[] = array('value' => $text, 'analyzer' => $analyzer);
        return $this;
    }


    public function getTotal()
    {
        return $this->total;
    }

    public function getAnalyzedWords()
    {
        return $this->analyzed;
    }

    public function getDocuments($offset = 0, $limit = 20)
    {
        $this->offset = $offset;
        $this->limit = $limit;

        $toJSON = array(
            'action' => 'search',
            'setType' => $this->type,
            'limitToCollection' => $this->collection,
            'limitToField' => $this->field,
            'analyzeText' => $this->analyze,
            'getDocuments' => array('offset' => $this->offset, 'limit' => $this->limit)
        );
        $result = $this->index->postRequest(json_encode($toJSON));
        $this->total = $result['total'];
        $this->analyzed = $result['analyzed'];

        foreach($result['documents'] as $k => $document_array)
        {
            $document = new Document($this->index);
            $document->setCollection($document_array['collection']);
            $document->setIdentifier($document_array['identifier']);
            foreach($document_array['metas'] as $meta_name => $meta_value)
            {
                $document->setMeta($meta_name, $meta_value);
            }
            $result['documents'][$k] = $document;
        }

        return $result['documents'];
    }
}


class Exception extends \Exception
{
    private $last_raw_response;
    public function setLastRawResponse($last_response)
    {
        $this->last_raw_response = $last_response;
    }

    public function getLastRawResponse()
    {
        return $this->last_raw_response;
    }
}